plaid/src/data/websocket/selector.rs

```diff
@@ -0,0 +1,124 @@
+    time::{Duration, Instant},
+};
+
+/// Represents a single URI entry with backoff duration and next attempt time.
+///
+/// The `UriEntry` struct contains information about a URI, including the duration to wait before
+/// retrying a connection (`backoff_duration`) and the time when the next connection attempt
+/// should be made.
+#[derive(Debug)]
+struct UriEntry {
+    name: String,
+    /// The URI to connect to.
+    uri: Uri,
+    /// The duration to wait before retrying a connection.
+    backoff_duration: Duration,
+    /// The time when the next connection attempt should be made.
+    next_attempt: Instant,
+}
+
+/// Manages a list of URI entries and handles the selection and retry logic for connection attempts.
+///
+/// The `UriSelector` struct contains a collection of `UriEntry` instances and manages the logic
+/// for selecting the next URI to attempt a connection to, including backoff and retry mechanisms.
+#[derive(Debug)]
+pub struct UriSelector {
+    /// A list of `UriEntry` instances.
+    uris: Vec<UriEntry>,
+    /// The index of the currently selected URI.
+    current_index: usize,
+    /// The initial duration to wait before retrying a connection.
+    initial_retry_after: Duration,
+    /// The maximum duration to wait before retrying a connection.
+    max_retry_after: Duration,
+}
+
+impl UriSelector {
+    /// Creates a new `UriSelector` with the given list of URIs, initial retry duration, and maximum retry duration.
+    ///
+    /// # Arguments
+    ///
+    /// * `uris` - A vector of `Uri` objects representing the URIs to manage.
+    /// * `initial_retry_after` - A `Duration` specifying the initial time to wait before retrying a failed URI.
+    /// * `max_retry_after` - A `Duration` specifying the maximum time to wait before retrying a failed URI.
+    ///
+    /// # Returns
+    ///
+    /// A `UriSelector` instance with the given URIs and retry durations.
+    pub fn new(
+        uris: HashMap<String, Uri>,
+        initial_retry_after: Duration,
+        max_retry_after: Duration,
+    ) -> Self {
+        let now = Instant::now();
+        UriSelector {
+            uris: uris
+                .into_iter()
+                .map(|(name, uri)| UriEntry {
+                    name,
+                    uri,
+                    backoff_duration: initial_retry_after,
+                    next_attempt: now,
+                })
+                .collect(),
+            current_index: 0,
+            initial_retry_after,
+            max_retry_after,
+        }
+    }
+
+    /// Selects the next URI to use, prioritizing URIs with the shortest backoff duration.
+    ///
+    /// This function sorts the URIs by their next attempt time,
+    /// and returns a tuple containing the name and URI of the entry with the shortest backoff duration that is ready for the next attempt.
+    /// If no URIs are ready, it selects and returns the name and URI of the entry with the earliest next attempt time.
```

Comment:
If this is the case then I think the concept of "Ready" perhaps is not the right word to use. I think we either need to stop the generator until one is ready, or change the wording to be that we will do exponential back off but it won't mean much if they all go down (since we're just going to cycle through them at high speed).

Reply:
Good call - implemented waiting until next attempt has passed